### ISSUE TITLE
Seems to not work for markdown files

### ISSUE THREAD

**Issue by baroldgene** [aaronik/treewalker.nvim#2]
Not sure if it's a facet of my markdown LSP or not, but this doesn't jump at all when browsing markdown files (specifically my obsidian vault).  Jumping to peers (`Treewalker Left` or `Treewalker Right`) simply goes up/down one line, even when on on things like headers.  Jumping to parent/children (`Treewalker Up` or `Treewalker Down`) doesn't move at all.  

Inspecting the Treesitter node tree it seems there are `section` nodes that encompass the `header` nodes which might be part of the problem.  See screenshot below.

<img width="673" alt="Screenshot 2024-12-12 at 12 00 18 PM" src="https://github.com/user-attachments/assets/650de386-713f-477f-9baa-0b5922fdbc48" />

**Comment by baroldgene** [comment]
Still an amazing plugin!  I'm super excited to use it if I can get the markdown thing solved!  (Sorry, meant to lead with that.)

**Comment by olimorris** [comment]
This will be because it's likely most markdown files will have two parsers associated with it: `markdown` and `markdown_inline`

**Comment by baroldgene** [comment]
I think I figured out what's going on.  I misunderstood what the plugin does.  To find candidate rows it doesn't actually traverse Treesitter, it just looks at the column of the text.  Since markdown isn't auto-indented it errantly grabs rows that aren't in the right treesitter context.  

It similarly breaks if you undo the formatting of a javascript file.  If I undo a few spaces in front of a function name it similarly won't be found as a child anymore.  Conversely if I add actual tabs for things like headers then it will jump properly.

**Comment by baroldgene** [comment]
@aaronik I mixed up what I said in the comment, but the behavior is still not working as expected.  This is due, in my troubleshooting, to the fact that my markdown files aren't actually tabbed.  For example, this markdown won't allow you to jump properly:
```markdown
# Header 1
## SubHeader 1
- List Item 1
- List Item 2
# Header 2
## SubHeader 2
- List Item 1
- List Item 2
```

However this markdown **will** work properly
```markdown
# Header 1
  ## SubHeader 1
    - List Item 1
    - List Item 2
# Header 2
  ## SubHeader 2
    - List Item 1
    - List Item 2
``` 

This holds true regardless of markdown or not, though.  If you intentionally remove spaces/tabs from a javascript file the jumping behavior will also cease to function properly.

**Comment by baroldgene** [comment]
Yeah sorry, definitely didn't mean it as a criticism.  Still a cool plugin, I'm just sad I can't really use it (until I find a workaround).

**Comment by aaronik** [comment]
Yeah man sorry, wish it could cover your use case. 

In the reddit post (https://www.reddit.com/r/neovim/s/XUSMyrsDwT), a few commenters have asked why I didn't just do such and such. Maybe one of those such and such's could work for you?

Or, even, maybe if your only use case is for markdown files, you could use Left (which I suspect may work for you still), Up, Down, and in place of Right, you could actually make your own child finding situation, like

```lua
function Down()
  local tw = require "treewalker"
  local current = vim.treesitter.get_node()
  local kid = tw.get_children(current)[1]
  if kid then
    tw.ops.jump(kid)
  end
end
```

I just wrote that on my phone, so barring any typos, is that something that might work for you?

The problem is some ASTs don't arrange themselves favorably to jump to a different starting point, then get_node gets the wrong node, and you get stuck. So it's either statefulness (which is a hard pass for me), or you need to start looking deeper down the descendant tree. I can help you with that if you want. I can't remember but there may even be a function in treewalker.nodes that does that

**Comment by baroldgene** [comment]
Yeah I was toying around with some of the treesitter API.  I was about to ask if you had already tried kinda traversing the tree in some way rather than using the indentation because I didn't want to waste my time if you'd already gone down that dead end.  

My thought was for siblings to traverse up the tree checking all the children to see if any of them matched the same child level and were in a list of approved node types and see where that got me.  The logic there is rough though and I'm still not clear that it will work.

**Comment by baroldgene** [comment]
Yeah I'm noticing that same thing.  Even trying to go by named nodes seems a waste as that's incredibly hit or miss.  

I do wonder if a list of white-listed nodes per language could be started and grown over time.  Also could be a config option for users to override if needed.  

I may toy with it a bit, but I'm not sure how far I'll get.  I'm relatively new to plugin dev and I haven't written code as a job in quite a while.

**Comment by aaronik** [comment]
Yeah you can play with it if you want but I bet you discover All the same issues I did. At one point I even tried to compute a new tree, hoping to save all the issues I was having with the original tree. That turned out to be a bust as well, very difficult algorithms in there. 

Is it possible to actually indent your markdown? I usually wouldn't change my code for the sake of a plugin, but if this type of movement meant that much to you, well it's at least a consideration.

Otherwise if you really do want to write some code here, I'd suggest:

Get the first node that is both:
* a descendant of the starting node, and
* is on a different line from the starting node

Hmmm I'm wondering if I ever tried that? I'm gonna see if I could actually do that here and still pass the tests

**Comment by baroldgene** [comment]
Yeah I was thinking of trying something along those lines.  My rough thought was:

move out: Traverse up the tree until I get an object that has a different line number to start

Move to sibling: move up the tree until I find a node that has a child of the same node type with the same depth and jump to that node

Moving in seems a bit more complicated.  Not sure to determine how deep down the tree to go.  Maybe first child that is on a new line?  Also wondering how well the logic for moving across siblings would work.  I don't work with enough languages to know if "siblings" all share the same type in most of their parsers.

**Comment by baroldgene** [comment]
Oh and I am also looking into adding tabs to markdown.  Not entirely sure the best way to auto-indent that really but I'd actually like it for other reasons.  One of the markdown plugins already displays it that way so it wouldn't bother me to have the file actually saved like that.  Need to research more.

**Comment by aaronik** [comment]
Ok so I tried this technique of getting the first descendant on a diff line, but it does not work for all files. Look at this tree, see, the problem is the current node is "name", but it has no children! That's how this tree is laid out.

<img width="1259" alt="Screenshot 2024-12-12 at 5 42 07 PM" src="https://github.com/user-attachments/assets/ef80e165-ade1-4290-9115-d85326f115a0" />

So now it's like, ok, maybe not just all descendants, but all descendants of current node and all of its siblings. 

But then that doesn't always work, so it's like Ok, then we need to traverse up the tree and get all descendants of all aunties and uncles? But then you start getting unwanted nodes in certain languages.

So here's the dilemma. This is why I resorted to going line by line. Because ultimately, when you're looking at code, no matter how the AST gets constructed, you're always wanting the same thing from a plugin like this. To go down and over.

**Comment by baroldgene** [comment]
Ok so I _think_ that I have a way that could work.  I cobbled together something to test the concept and it seems to work for the few examples I've tried it on (JS, Markdown).  Since the jumping is more line based than column based I essentially combine multiple node levels into one based on the starting line number.

The rough strategy is: 

For Up/Down: 
* Traverse up to the first node that has a different line number than the current node
* Look at all children of that node and jump to the first one with a line number that's < or > the line number of the starting node
* Jump to that line

For Left:
* Jump to the first parent with a different line number

For Right (haven't actually done this one yet:
* Jump to the first child with a different line number.


I was gonna try to flesh this out so that the plugin actually works (so far I've just cobbled together some methods to test).  My question for you is: This would be a pretty heavy refactor and arguably an entirely different plugin.  I wasn't sure if I should make a fork, a new plugin, or a PR.  I don't want to steal your work (which is pretty epic) but also didn't want to overhaul it too extensively if it was heading in a direction that didn't fit your use case.

**Comment by aaronik** [comment]
I say implement your way and check back with me when the tests are passing

**Comment by bpstahlman** [comment]
I read through the posts but didn't really understand the motivation for making navigation depend on line/column numbers. One of the nice things about Treesitter, and ASTs more generally, is that they enable consistent code navigation, without regard to something as variable and insignificant as whitespace formatting. The AST parser cares no more about whitespace than does the language it's parsing, so why should a plugin that walks the AST be any different? Taking the following Lua snippet as an example...
```
<1>{ mode = { "o" },           'r',     function() require("flash").remote() end },    
<2>{
  mode = { "n", "x", "o" },
  '<A-t>',
  function() require("flash").treesitter() end
},                                                                                                                                                                                                                                                                                                                                                                             
```
With my cursor on node `<1>`, `Treewalker Right` takes me to the first child of node `<2>`. This feels broken, given that node `<1>` has its own child nodes. Why should it matter that the children are on the same line as the parent? I was very excited to discover this plugin, but now I'm very confused.

**Comment by baroldgene** [comment]
@bpstahlman I thought the same until I started diving into this a bit more.  As noted the issue with treesitter and lsps in general is that they don't follow a standard convention which leads to a ton of variation.  This makes it incredibly hard to reliably know the proper amount of jumps up/down the tree to start considering siblings.  A quick workaround for this is to use the indentation since that's how the user is viewing the code anyway.  

I was looking at trying to use solely the tree-sitter parser for a bit and I do think it might be possible, but I got pulled into other things and haven't had a chance to revisit it.  And tbh given that @aaronik is more knowledgeable than I am and didn't solve it I think it's fairly likely that it really isn't possible to do this in a clean way.

**Comment by bpstahlman** [comment]
@baroldgene Hmm... Just played around with the Treesitter API in a Lua buffer and I don't see the issue. For the Lua code example I gave, the Treesitter node structures look ***identical*** (other than line/col numbers) whether the node spans one line or many lines. I would submit that a Treesitter parser that _didn't_ produce identical tree structures for those two cases would be broken. The builtin commands used to navigate the `:InspectTree` buffer have no trouble navigating to children on the same line as the parent. If you hit `a`, it will also treat syntax elements such as `{` and `=` as navigable children, but these correspond to "anonymous" nodes and could be easily skipped by a tree-navigating plugin. The point is that for Treesitter, the only thing distinguishing something like...
```
{
  1,
  2,
  3
}
```
...from `{1,2,3}` is the range (line/col) of the nodes. What am I missing?

**Comment by baroldgene** [comment]
The issue isn't walking the tree in one language.  That's easy.  The issue is making code to walk the tree for, say Lua and having that same code also work for Rust, or Go, or Markdown.  

I would have to dig back in to provide specifics but I recall trying to work on this using Lua and Markdown as my two examples and couldn't get something working that would handle both.  (I will say that the treesitter parser for markdown is "broken" as you said IMO though.  Or at least it operates differently from any other language parse I could find)

Being honest though I'm not very familiar with Treesitter so maybe it's something obvious that I don't understand.  If you have a solution on how to make it work you could submit a PR?  I'm happy to collaborate on the PR if you'd like, not that I'd be terribly useful.

**Comment by bpstahlman** [comment]
After reading the above, I used `:InspectTree` on both a Rust and a Markdown file, using the builtin maps to navigate each. In both cases, the navigation was seamless. Moreover, the elements of single-line data structures were navigated normally. I watched the Treesitter buffer while splitting a single-line Rust data structure into multiple lines and verified that **nothing** about the tree structure changed except for the line/col numbers. As for Markdown, the `section` nodes you referred to seem to be just wrapper nodes for an entire section, so I'm not sure how they would cause a problem.

If I don't decide on a different tree navigation plugin first, I'll probably delve into the source code to see what's going on. I could still be overlooking something, but at this point, it seems as though the whitespace-aware logic added to fix some sort of corner case has completely broken what should be (IMO) the primary use case: navigating source code. Will let you know if I find out otherwise and/or end up going the PR route...

**Comment by baroldgene** [comment]
@bpstahlman I'm definitely not saying you're wrong, just saying I wasn't able to figure out how to reliably traverse the nodes as expected.  If you can outline a high level strategy for what you're thinking I can try to code it up (either in a PR or a fork).  I was at it for a bit but then got pulled away by stuff at my actual job.

**Comment by bpstahlman** [comment]
> [@bpstahlman](https://github.com/bpstahlman) I'm definitely not saying you're wrong, just saying I wasn't able to figure out how to reliably traverse the nodes as expected. If you can outline a high level strategy for what you're thinking I can try to code it up (either in a PR or a fork). I was at it for a bit but then got pulled away by stuff at my actual job.

Ha, ha. I definitely know how that is... But I've been digging into nvim scripting lately as part of my transition from Vim to Neovim, so it may be a good exercise.

**Comment by bpstahlman** [comment]
> Tree-climber is going to be your thing. It treats all nodes literally.

@aaronik Thanks! I'll give it a try...
